contracts/exchange-forwarder/contracts/src/libs/LibAssetDataTransfer.sol

```diff
@@ -0,0 +1,258 @@
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+
+*/
+
+pragma solidity ^0.5.9;
+pragma experimental ABIEncoderV2;
+
+import "@0x/contracts-utils/contracts/src/LibBytes.sol";
+import "@0x/contracts-utils/contracts/src/LibRichErrors.sol";
+import "@0x/contracts-utils/contracts/src/LibSafeMath.sol";
+import "@0x/contracts-erc20/contracts/src/LibERC20Token.sol";
+import "@0x/contracts-erc721/contracts/src/interfaces/IERC721Token.sol";
+import "@0x/contracts-erc1155/contracts/src/interfaces/IERC1155.sol";
+import "@0x/contracts-asset-proxy/contracts/src/interfaces/IAssetData.sol";
+import "./LibForwarderRichErrors.sol";
+
+
+library LibAssetDataTransfer {
+
+    using LibBytes for bytes;
+    using LibSafeMath for uint256;
+    using LibAssetDataTransfer for bytes;
+
+    /// @dev Transfers given amount of asset to sender.
+    /// @param assetData Byte array encoded for the respective asset proxy.
+    /// @param from Address to transfer asset from.
+    /// @param to Address to transfer asset to.
+    /// @param amount Amount of asset to transfer to sender.
+    function transferFrom(
+        bytes memory assetData,
+        address from,
+        address to,
+        uint256 amount
+    )
+        internal
+    {
+        if (amount == 0) {
+            return;
+        }
+
+        bytes4 proxyId = assetData.readBytes4(0);
+
+        if (
+            proxyId == IAssetData(address(0)).ERC20Token.selector ||
+            proxyId == IAssetData(address(0)).ERC20Bridge.selector
+        ) {
+            assetData._transferERC20Token(
+                from,
+                to,
+                amount
+            );
+        } else if (proxyId == IAssetData(address(0)).ERC721Token.selector) {
+            assetData._transferERC721Token(
+                from,
+                to,
+                amount
+            );
+        } else if (proxyId == IAssetData(address(0)).ERC1155Assets.selector) {
+            assetData._transferERC1155Assets(
+                from,
+                to,
+                amount
+            );
+        } else if (proxyId == IAssetData(address(0)).MultiAsset.selector) {
+            assetData._transferMultiAsset(
+                from,
+                to,
+                amount
+            );
+        } else if (proxyId != IAssetData(address(0)).StaticCall.selector) {
+            LibRichErrors.rrevert(LibForwarderRichErrors.UnsupportedAssetProxyError(
+                proxyId
+            ));
+        }
+    }
+
+    ///@dev Transfer asset from sender to this contract.
+    /// @param assetData Byte array encoded for the respective asset proxy.
+    /// @param amount Amount of asset to transfer to sender.
+    function transferIn(
```

Comment:
Do we use this? Is this in anticipation of property-based orders?

Reply:
Yep, this is in anticipation of property based orders and arbitrary token sell support. We do not use this yet.